<center>
    <img src="https://raw.githubusercontent.com/dvgodoy/PyTorch101_ODSC_Europe2020/master/images/linear_dogs.jpg" width="800"> 
</center>  

# Yet another math for DS course: домашнее задание 5

**ФИО:**

## Формат сдачи

Сам ноутбук называйте в формате hw-07-USERNAME.ipynb, где USERNAME — ваши фамилия и имя. Файл надо будет сдать в anytask.

## Задача №1: точность вычислений

С точки зрения математики матричные разложения являются точными: произведение сомножителей всегда равняется исходной матрицы $A$. К сожалению, на практике этом часто мешает вычислительная погрешность. 

Для $LU$ разложения $l_2$-норма ошибки ошибки $||\delta A|| = ||A - LU||$ удовлетворяет следующей оценке:

$$||\delta A|| \leqslant ||L|| \cdot ||U|| \cdot O(\varepsilon_{machine})$$

А нормы $L$ и $U$ могут быть совсем нехорошими.

__[2 балла] Задание 1.1__ Рассмотрим следующее LU-разложение:

$$
\begin{pmatrix}
1 & 0\\
10^{20} & 1
\end{pmatrix}
\cdot
\begin{pmatrix}
10^{-20} & 1\\
0 & 1 - 10^{20}
\end{pmatrix}
$$
и
$$
\begin{pmatrix}
1. & 0.\\
10.^{20} & 1.
\end{pmatrix}
\cdot
\begin{pmatrix}
10.^{-20} & 1.\\
0. & 1. - 10.^{20}
\end{pmatrix}
$$

Во втором случае мы специально указали Питону (поставив везде десятичные точки), что хотим работать не с целыми числами, а с числами с плавающей точкой. Перемножьте полученные матрицы $L$ и $U$. Как ответ вам кажется правильным, почему?

> __Ответ:__ В первом случае идет точное вычисление элементов матрицы, так как мы используем для вычисления Python, а не сам процессор. В результате получается точно значение, которое не моделирует реальный процесс вычисления. То есть фактически это правильный ответ, но не соответствует реальности. Поэтому второй случай правильный. 
Также стоит отметить, что LU имеют большие нормы (1e+20, 1e+20), что говорит о большой вычислительной ошибки, если использовать LU разложение. Такие матрицы лучше не использовать для LU разложения, либо скорректировать подход. 


In [88]:
import numpy as np 


L = np.array([[1, 0], [10**20, 1]])
U = np.array([[10**-20, 1], [0, 1 - 10**20]])
A = L @ U


A


array([[1e-20, 1],
       [1.0, 1]], dtype=object)

In [96]:
np.linalg.norm(U), np.linalg.norm(A, 'fro')

(np.float64(1e+20), np.float64(1.7320508075688772))

In [93]:
L_ = np.array([[1., 0.], [10.**20, 1.]])
U_ = np.array([[10.**-20, 1.], [0., 1. - 10.**20]])
A_ = L_ @ U_

A_

array([[1.e-20, 1.e+00],
       [1.e+00, 0.e+00]])

In [94]:
np.linalg.cond(A_)

np.float64(1.0)

In [95]:
np.linalg.norm(L_, 'fro'), np.linalg.norm(U_, 'fro'), np.linalg.norm(A_, 'fro')

(np.float64(1e+20), np.float64(1e+20), np.float64(1.4142135623730951))

In [ ]:
#Эксперимент 
import scipy as sc 

sc.linalg.lu(A_)

(array([[0., 1.],
        [1., 0.]]),
 array([[1.e+00, 0.e+00],
        [1.e-20, 1.e+00]]),
 array([[1., 0.],
        [0., 1.]]))

Отметим, что в реальных вычислениях матричные элементы почти наверняка с самого начала будут числами с плавающей точкой (а не целыми).

**Выход: LU-разложение с выбором главного элемента (по столбцу)**

Каждый раз ищем максимум в столбце и переставляем соответствующую строку наверх.

$$\begin{pmatrix}
b_{11} & \dots & b_{1i} & b_{1,i+1} & \dots & b_{1n}\\
 & \ddots & \vdots & \vdots & & \vdots\\
 & & \color{blue}{b_{ii}} & \color{blue}{b_{i,i+1}} & \dots & \color{blue}{b_{in}} \\
 & & b_{i+1,i} & b_{i+1,i+1} & \dots & b_{i+1,n}\\
 & & \vdots & \vdots &  & \vdots \\
 & & \color{green}{b_{ji}} & \color{green}{b_{j,i+1}} & \dots & \color{green}{b_{jn}} \\
 & & \vdots & \vdots & & \vdots\\
\end{pmatrix}\longrightarrow
\begin{pmatrix}
b_{11} & \dots & b_{1i} & b_{1,i+1} & \dots & b_{1n}\\
 & \ddots & \vdots & \vdots & & \vdots\\
 & & \color{green}{b_{ji}} & \color{green}{b_{j,i+1}} & \dots & \color{green}{b_{jn}} \\
 & & b_{i+1,i} & b_{i+1,i+1} & \dots & b_{i+1,n}\\
 & & \vdots & \vdots &  & \vdots \\
 & & \color{blue}{b_{ii}} & \color{blue}{b_{i,i+1}} & \dots & \color{blue}{b_{in}} \\
 & & \vdots & \vdots & & \vdots\\
\end{pmatrix}\longrightarrow$$
$$\longrightarrow\begin{pmatrix}
b_{11} & \dots & b_{1i} & b_{1,i+1} & \dots & b_{1n}\\
 & \ddots & \vdots & \vdots & & \vdots\\
 & & \color{green}{b_{ji}} & \color{green}{b_{j,i+1}} & \dots & \color{green}{b_{jn}} \\
 & & 0 & b'_{i+1,i+1} & \dots & b'_{i+1,n}\\
 & & \vdots & \vdots &  & \vdots \\
 & & 0 & b'_{i,i+1} & \dots & b'_{in} \\
 & & \vdots & \vdots & & \vdots
\end{pmatrix}$$

Надо сказать, что примерно так вы все и решали системы на первом курсе университета! Именно наибольший, а не первый ненулевой элемент столбца берётся потому, что чем больше число - тем меньшие погрешности потенциально вносит деление на него.

Что при этом происходит? Перестановка строк матрицы равносильна умножению её слева на матрицу соответствующей перестановки. Таким образом, мы получаем равенство

$$L_nP_nL_{n-1}P_{n-1}\ldots L_2P_2L_1P_1 A = U\qquad\qquad(1)$$

где $L_1,\ldots,L_n$ - некоторые нижнетреугольные матрицы.

**Вопрос:** Ну, и где здесь матрица $L$?!

**Ответ:** Введём новые матрицы

\begin{align*}
L'_n &= L_n\\
L'_{n-1} &= P_nL_nP_{n-1}\\
L'_{n-2} &= P_nP_{n-1}L_{n-1}P_n^{-1}P_{n-1}^{-1}\\
&\ldots\\
L'_1 &= P_nP_{n-1}\ldots P_2L_1P_2^{-1}\ldots P_{n-1}^{-1}P_n^{-1}
\end{align*}

**Упражнение.** Матрицы $L'_i$ тоже нижнетреугольные!

Тогда левая часть (1) перепишется в виде

$$\underbrace{L'_nL'_{n-1}\ldots L'_1}_{:=L^{-1}}\underbrace{P_nP_{n-1}\ldots P_1}_{:=P^{-1}}\cdot A$$

**Итог:** разложение вида
$$A = PLU$$
где $P$ - матрица перестановки.

Функция `scipy.linalg.lu` в Питоне находит именно такое разложение!

Все элементы $L$ не превосходят $1$, так что $||L||]\leqslant 1$. При этом
$$||\Delta A|| \leqslant ||A||\cdot O(\rho \varepsilon_{machine}),$$
где
$$\rho = \frac{\max_{i,j}|u_{ij}|}{\max_{i,j}|a_{ij}|}$$
Это число называется *фактором роста матрицы*.

Но что, если это отношение велико?

__[2 балла] Задание 1.2__  Сгенерируйте матрицу $500\times500$, имеющую вид

$$\begin{pmatrix}
1 & 0 & 0 & \cdots & 0 & 0 & 1\\
-1 & 1 & 0 &  &  & 0 & 1\\
-1 & -1 & 1 & 0  &  & 0 & 1\\
\vdots & & \ddots & \ddots  & \ddots & \vdots & \vdots \\
-1 & -1 & -1 & \ddots & 1 & 0 & 1\\
-1 & -1 & -1 &  & -1 & 1 & 1\\
-1 & -1 & -1 & \cdots & -1 & -1 & 1
\end{pmatrix}$$

Например, вы можете сгенерировать сначала нулевую матрицу нужного размера, а потом заполнить её клетки правильными числами.

Найдите её PLU-разложение и QR-разложение. Убедитесь, что $P = E$. Вычислите $||A - LU||_2$ и $||A - QR||_2$. Чему равен фактор роста матрицы $A$?

In [205]:
# your code here
# (｡❤‿❤｡)
import scipy  as sc 

A = np.zeros((500, 500))
np.fill_diagonal(A, 1)
A += -np.tril(np.ones((500, 500), dtype=int), k=-1)
A[:, -1] = 1



Amax = np.max(np.abs(A))

##### Разложение LU

In [206]:
A_PLU = sc.linalg.lu(A)

In [207]:
U = A_PLU[2]
L = A_PLU[1]


In [208]:
np.linalg.norm((A-L@U), 2) # норма ошибки LU

np.float64(3.102517070422723e+116)

In [209]:
A_Umax = np.max(np.abs(A_PLU[2]))
A_U_max / Amax 



np.float64(1.636695303948071e+150)

##### Разложение QR

In [210]:
A_QR = sc.linalg.qr(A)
Q = A_QR[0]
R = A_QR[1]

In [213]:
np.linalg.norm((A-Q@R), 2)

np.float64(1.0872138111957882e-12)

##### Фактор роста  при PLU разложении ~ 1.636695303948071e+150


К счастью, на практике так редко очень бывает (чёрт его знает почему). Тем не менее, QR-разложение всё-таки лучше. Теоретическая оценка для ошибки QR-разложения имеет вид

$$||A - QR||_2 \leqslant ||A||_2\cdot O(\varepsilon_{machine})$$

__[2 балла] Задание 1.3__  Рассмотрим *матрицу Паскаля* $S_n = \left(C_{i + j}^i\right)$ ($i,j = 0,\ldots,n-1$).

Каково её LU-разложение? Выведите формулы для матриц L и U и приведите краткое обоснование прямо в ноутбуке. Не пользуйтесь функцией `scipy.linalg.lu`, чтобы его "угадать": матрица P будет отлична от единичной, и вы получите не то, что хотели.

Каков её определитель?

> __Ответ:__ 

__[4 балла] Задание 1.4__ Напишите функцию `my_pascal(n)`, генерирующую матрицу Паскаля размера $n\times n$.

In [249]:
def element(i, j)-> int: # Расчет элемента симметриченой матрица Паскаля 
    n = i + j 
    r = i 

    el = factorial(n)/(factorial(r)*factorial(n-r))

    return el 


In [250]:
import numpy as np 
from math import factorial

def my_pascal(n):   # 
    A = np.zeros((n,n))
    
    for i in range(n):
        for j in range(n):
            A[i, j] = element(i, j)
            
    return A 






    

Найдите норму разности $||A - PLU||_2$. Не такая уж и большая, правда?

In [323]:

# your code here
# (❍ᴥ❍ʋ)
# Зададим матрицу паскаля 30x 30 \
n = 30 
A = my_pascal(n)

P, L, U = sc.linalg.lu(A)

A_LU = P @ L @ U

sc.linalg.norm((A-A_LU), 2)

np.float64(4.808165824568957)

Теперь попросим компьютер вычислить определитель матрицы Паскаля $30\times30$ и решить простенькую систему уравнений:

In [ ]:
# your code here
#  ლ(ಠ益ಠლ)
# 

det_A_PLU = np.linalg.det(P) * np.prod(np.diag(U))
det_A = np.linalg.det(A)

det_A_PLU, det_A






(np.float64(1.0299188904699096e+48), np.float64(1.0299188904698675e+48))

In [326]:
b = 2 ** np.arange(n)

y = np.linalg.solve(L, P.T @ b)
x_lu = np.linalg.solve(U, y)
x_direct = np.linalg.solve(A, b)


norma_lu = np.linalg.norm(x_lu - x_direct)
norma_lu

np.float64(0.0)

Так себе ошибка. Теперь попробуем сделать это с помощью QR-разложения. Станет ли лучше?

In [327]:
# your code here
# (⌐■_■)



Q, R = np.linalg.qr(A)
x_qr = np.linalg.solve(R, Q.T @ b)
norma_qr = np.linalg.norm(x_qr - x_direct)

norma_qr 


np.float64(1727649.735079213)

Объясните полученные неприятные результаты.

> __Ответ:__ Ошибка в случае решения через LU разложение матрицы Паскаля 30 на 30 меньше, чем через QR разложение. Скорее всегоо это связано с тем, что в LU у нас просто перестановки, а в QR разложении присутствует нормировка и из за больних чисел сильно влияет на ошибку 